src/dbnode/storage/index/compaction/plan.go

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package compaction
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/storage/index/segments"
+)
+
+var (
+	errMutableCompactionAgeNegative = errors.New("mutable compaction age must be postive")
+	errSizeBucketsUndefined         = errors.New("size buckets are undefined")
+	errMaxImmutableCompactionSize   = errors.New("max immutable compaction size must be positive")
+)
+
+var (
+	DefaultSizeBuckets = []SizeBucket{ // i.e. tiers for compaction [0, 262K), [262K, 524K), [524K, 1M), [1M, 2M), [2M, 8M)
+		SizeBucket{
+			MinSizeInclusive: 0,
+			MaxSizeExclusive: 1 << 18,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 18,
+			MaxSizeExclusive: 1 << 19,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 19,
+			MaxSizeExclusive: 1 << 20,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 20,
+			MaxSizeExclusive: 1 << 21,
+		},
+		SizeBucket{
+			MinSizeInclusive: 1 << 21,
+			MaxSizeExclusive: 1 << 23,
+		},
+	}
+
+	DefaultOptions = PlannerOptions{
+		MaxImmutableCompactionSize: 1 << 22,                            // ~4M
```

Comment:
Hm, shouldn't this just be computed from the size buckets?  Doesn't seem to buy too much by having another max size other than the upper limit of the largest size bucket?  I see that this is a value in the last default bucket itself, but couldn't you just set the last bucket to 2m to 4m and then not need this number at all?

Reply:
As things are, it's definitely specifying the same thing as controlling the last bucket size. The original intention with the knob was to have an override which ensures we don't compact things that are too large, regardless of the bucket size. It'd be useful in cases where we want to do a non-leveled compaction (e.g. once we know a block is sealed, we want to compact as much as possible).

Happy to revert for now if you'd prefer, can add back in once we add the different planner types.